pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,215 @@
+}
+
+// addSecretEventHandler should only be used directly for tests. For production use AddSecretEventHandler().
+// createInformerFn helps in mocking sharedInformer for unit tests.
+func (s *secretMonitor) addSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler, createInformerFn func() cache.SharedInformer) (SecretEventHandlerRegistration, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !validateName(routeSecretName) {
+		return nil, fmt.Errorf("invalid routeSecretName combination : %s", routeSecretName)
+	}
+
+	// Inside a namespace multiple route can refer a common secret.
+	// Hence route1_secret and route2_secret should denote separate key,
+	// so separate monitors inside a namespace.
+	key := NewObjectKey(namespace, routeSecretName)
+
+	// Check if monitor / watch already exists
+	m, exists := s.monitors[key]
+	if !exists {
+		secretName, err := getSecretName(routeSecretName, true)
+		if err != nil {
+			return nil, err
+		}
+
+		var sharedInformer cache.SharedInformer
+		if createInformerFn == nil {
+			// create a single secret monitor
+			sharedInformer = cache.NewSharedInformer(
+				cache.NewListWatchFromClient(
+					s.kubeClient.CoreV1().RESTClient(),
+					"secrets",
+					namespace,
+					fields.OneTermEqualSelector("metadata.name", secretName),
+				),
+				&corev1.Secret{},
+				0,
+			)
+		} else {
+			// only for testability
+			klog.Warning("creating informer for testability")
+			sharedInformer = createInformerFn()
+		}
+
+		m = newSingleItemMonitor(key, sharedInformer)
+		go m.StartInformer()
+
+		// add item key to monitors map // add watch to the list
+		s.monitors[key] = m
+
+		klog.Info("secret informer started", " item key ", key)
+	}
+
+	klog.Info("secret handler added", " item key ", key)
+
+	return m.AddEventHandler(handler) // also populate key inside handlerRegistartion
+}
+
+// Remove secret watch
+func (s *secretMonitor) RemoveSecretEventHandler(handlerRegistration SecretEventHandlerRegistration) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if handlerRegistration == nil {
+		return fmt.Errorf("secret handler is nil")
+	}
+	key := handlerRegistration.GetKey()
+
+	// check if watch already exists for key
+	m, exists := s.monitors[key]
+	if !exists {
+		klog.Info("secret monitor already removed", " item key", key)
+		return nil
+	}
+
+	if err := m.RemoveEventHandler(handlerRegistration); err != nil {
+		return err
+	}
+	klog.Info("secret handler removed", " item key", key)
+
+	// stop informer if there is no handler
+	if m.numHandlers.Load() <= 0 {
+		if !m.StopInformer() {
+			klog.Warning("secret informer already stopped", " item key", key)
+		}
+		delete(s.monitors, key)
+		klog.Info("secret informer stopped", " item key ", key)
+	}
+
+	return nil
+}
+
+// Get the secret object from informer's cache
+func (s *secretMonitor) GetSecret(handlerRegistration SecretEventHandlerRegistration) (*v1.Secret, error) {
```

Comment:
this function will work much more reliably for you if you can wait for the first sync when you add a new secret.

Reply:
Added wait after `go m.StartInformer()` as suggested in the next comment also. Let me know if this is what you are referring here also. 
```go
// wait for first sync
if !cache.WaitForCacheSync(context.Background().Done(), m.HasSynced) {
	return nil, fmt.Errorf("failed waiting for cache sync")
}
```